troposphere/static/js/components/modals/instance/InstanceLaunchWizardModal.react.js

```diff
@@ -1,396 +1,391 @@
     },
 
-    onRequestResources: function(){
-      // Launching a resource request modal will eat the current modal. We need to pass this.cancel as a prop
-      // in order to properly unmount the whole modal, not just the current step component.
-      this.cancel();
-      modals.HelpModals.requestMoreResources();
-    },
+    getInitialState: function() {
 
-    isLicenseStepActive: function() {
-      state = this.getState();
-      if (state.version
-        && state.version.get('licenses')
-        && state.version.get('licenses').length > 0) {
-        return true;
-      } else {
-        return false;
-      }
-    },
-    isOptionsStepActive: function() {
-      state = this.getState();
-      if (state.showOptions === true) {
-        return true;
-      } else {
-        return false;
-      }
+        // We might have these
+        let image = this.props.image ? this.props.image : null;
+        let imageName = image ? image.get("name") : null;
+        let project = this.props.project ? this.props.project : null;
+        let view = this.props.initialView;
+
+        return {
+            // State for general operation (switching views, etc) 
+            view,
+            image,
+            provider: null,
+
+            // State for launch
+            instanceName: null, // NOTE: this is only set when a user types in a specific name
+            imageVersion: null,
+            project,
+            providerSize: null,
+            identityProvider: null,
+            attachedScripts: [],
+        }
     },
-    renderImageSelect: function() {
-      return (
-                <ImageSelectStep
-                    image={this.state.image}
-                            onPrevious={this.cancel}
-                            onNext={this.onNext}
-                />
-      );
+
+    // Set the components state based on cloud defaults. 
+    //
+    // Whenever the wizard mounts it listens for changes from the stores,
+    // passing this function as a callback. Incrementally it calls stores to
+    // fetch data. It only sets state for defaults, i.e if project is null,
+    // set the project to the first returned from the cloud. It primes our
+    // stores, so that render can just call get and eventually get data.
+    updateState: function() {
+        var project = this.state.project;
+        if (!project) {
+            project = stores.ProjectStore.getAll().first();
+        }
+
+        var imageVersionList;
+        if (this.state.image) {
+            imageVersionList = stores.ImageVersionStore.fetchWhere({image_id: this.state.image.id});
+        }
+
+        var imageVersion = this.state.imageVersion;
+        if (imageVersionList && !imageVersion) {
+            imageVersion = imageVersionList.last();
+        }
+
+        var providerList, provider;
+        if (imageVersion) {
+            providerList = new Backbone.Collection(imageVersion.get('machines').map((item) => item.provider));
```

Comment:
I can look at making this a network request. The down side of course is adding another async call to the stack thus more complexity but with Connor's refactor this should be less so than it was.